packages/devtools_app/lib/src/timeline/flutter/timeline_screen.dart

```diff
@@ -151,14 +130,71 @@ class TimelineScreenBodyState extends State<TimelineScreenBody>
         child: Row(
           children: [
             Switch(
-              value: controller.timelineMode == TimelineMode.frameBased,
+              value: mode == TimelineMode.frameBased,
               onChanged: _onTimelineModeChanged,
             ),
             const Text('Show frames'),
           ],
         ),
       ),
     ];
+    if (mode == TimelineMode.frameBased) {
+      return ValueListenableBuilder(
```

Comment:
nice to see more of this code switch to ValueListenableBuilder instead of adhock change detection with no-op set state calls.

Reply:
I'm not sure this pattern really helps the code much.  The builders increase nesting, making the code harder to read.

Listening to a listener separately separates the concerns of UI layout from state changes, and the pattern of listening is straightforward: listen in initState() and cancel in dispose().  During build, it's fine to just retrieve `listener.value`.

This is especially more readable in cases where multiple listeners are used:

```dart
void initState() {
  void refresh() => setState((){}); // we can define this on our disposal mixin.
  addListener(listener1, refresh);
  addListener(listener2, refresh);
  addListener(listener3, refresh);
}

// We've already automated dispose.

Widget build(context) {
  return Text('${listener1.value}, ${listener2.value * listener3.value}');
}

```

Nesting ValueListenableBuilders to do the same thing is significantly more complicated.

Reply:
Discussed offline and pulled ValueListenableBuilder widgets out where necessary.